oak_idl/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! A Rust-based IDL implemented via `macro_rules`. It is inspired by
+//! <https://github.com/google/tarpc> and
+//! <https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format>.
+//!
+//! For each service definition, it generates both a client stub and a service trait, as well as a
+//! method to dispatch invocations to a service implementation.
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub enum ClientError {
+    InvalidRequest,
+    InvalidResponse,
+    TransportError(TransportError),
+}
+
+impl From<TransportError> for ClientError {
+    fn from(e: TransportError) -> Self {
+        ClientError::TransportError(e)
+    }
+}
+
+#[derive(Debug)]
+pub enum TransportError {
+    InvalidMessage,
+    InvalidMethodId,
+}
+
+/// Unique identifier of a method within a service.
+type MethodId = u32;
+
+/// Header of a transactional message.
+///
+/// See <https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format#transactional-messages>
+#[derive(Serialize, Deserialize)]
+pub struct Header {
+    pub transaction_id: u32,
+    pub method_id: MethodId,
+}
+
+/// A transactional message.
+///
+/// The body field contains the serialized payload of the request or response.
+///
+/// Note that the serialization of the body does not have to match the serialization of the Message
+/// itself. For instance, the Message may be serialized via bincode, but the paload may be
+/// serialized via protobuf.
+#[derive(Serialize, Deserialize)]
+pub struct Message {
+    pub header: Header,
+    pub body: Vec<u8>,
+}
+
+/// A message-oriented transport that allows performing invocations.
+pub trait Transport {
+    fn invoke(&self, request_message: &Message) -> Result<Message, TransportError>;
+}
+
+/// A wrapper for a message-oriented channel handle, which implements the [`Transport`] trait.
+struct Handle(u32);
+
+impl Transport for Handle {
+    fn invoke(&self, request_message: &Message) -> Result<Message, TransportError> {
+        // The serialization of the outer message is fixed by the transport, since the header needs
+        // to be deserializable by the other end of the transport, but the body of the
+        // request / response is serialized at a higher level, and in potentially different format,
+        // since that's handled by the application.
+        let request_bytes =
+            bincode::serialize(&request_message).map_err(|_| TransportError::InvalidMessage)?;
+        let response_bytes = invoke(self.0, &request_bytes)?;
+        let response_message =
+            bincode::deserialize(&response_bytes).map_err(|_| TransportError::InvalidMessage)?;
+        Ok(response_message)
```

Comment:
Seeing that invoke is transactional, I think it would be a good idea to validate that the transaction id of the response is the same as the one in the request and return an error if not. Possibly the same for the method id.